op-e2e/actions/proofs/holocene_batches_test.go

```diff
@@ -0,0 +1,144 @@
+				safeHeadHolocene:    0, // batch for block 2 is considered invalid because it is from the future. This batch + remaining channel is dropped.
+			},
+		},
+
+		{name: "case-2c", blocks: []uint{1, 1, 2, 3},
+			holoceneExpectations: holoceneExpectations{
+				safeHeadPreHolocene: 3, // duplicate batches are silently dropped, so this reduceds to case-0
+				safeHeadHolocene:    3, // duplicate batches are silently dropped
+			},
+		},
+		{name: "case-2d", blocks: []uint{2, 2, 1, 3},
+			holoceneExpectations: holoceneExpectations{
+				safeHeadPreHolocene: 3, // duplicate batches are silently dropped, so this reduces to case-2b
+				safeHeadHolocene:    0, // duplicate batches are silently dropped, so this reduces to case-2b
+			},
+		},
+	}
+
+	runHoloceneDerivationTest := func(gt *testing.T, testCfg *helpers.TestCfg[testCase]) {
+		t := actionsHelpers.NewDefaultTesting(gt)
+		env := helpers.NewL2FaultProofEnv(t, testCfg, helpers.NewTestParams(), helpers.NewBatcherCfg())
+
+		includeBatchTx := func() {
+			// Include the last transaction submitted by the batcher.
+			env.Miner.ActL1StartBlock(12)(t)
+			env.Miner.ActL1IncludeTxByHash(env.Batcher.LastSubmitted.Hash())(t)
+			env.Miner.ActL1EndBlock(t)
+
+			// Finalize the block with the first channel frame on L1.
+			env.Miner.ActL1SafeNext(t)
+			env.Miner.ActL1FinalizeNext(t)
+		}
+
+		var max = func(input []uint) uint {
+			max := uint(0)
+			for _, val := range input {
+				if val > max {
+					max = val
+				}
+			}
+			return max
+		}
+
+		targetHeadNumber := max(testCfg.Custom.blocks)
+		for env.Engine.L2Chain().CurrentBlock().Number.Uint64() < uint64(targetHeadNumber) {
+			env.Sequencer.ActL2StartBlock(t)
+			// Send an L2 tx
+			env.Alice.L2.ActResetTxOpts(t)
+			env.Alice.L2.ActSetTxToAddr(&env.Dp.Addresses.Bob)
+			env.Alice.L2.ActMakeTx(t)
+			env.Engine.ActL2IncludeTx(env.Alice.Address())(t)
+			env.Sequencer.ActL2EndBlock(t)
+		}
+
+		// Buffer the blocks in the batcher.
+		env.Batcher.ActCreateChannel(t, testCfg.Custom.isSpanBatch)
+		for _, blockNum := range testCfg.Custom.blocks {
+			env.Batcher.ActAddBlockByNumber(t, int64(blockNum), actionsHelpers.BlockLogger(t))
+		}
+		env.Batcher.ActL2ChannelClose(t)
+		frame := env.Batcher.ReadNextOutputFrame(t)
+		require.NotEmpty(t, frame)
+		env.Batcher.ActL2BatchSubmitRaw(t, frame)
+		includeBatchTx()
+
+		// Instruct the sequencer to derive the L2 chain from the data on L1 that the batcher just posted.
+		env.Sequencer.ActL1HeadSignal(t)
+		env.Sequencer.ActL2PipelineFull(t)
+
+		l2SafeHead := env.Sequencer.L2Safe()
+		testCfg.Custom.RequireExpectedProgress(t, l2SafeHead, testCfg.Hardfork.Precedence < helpers.Holocene.Precedence, env.Engine)
+
+		t.Log("Safe head progressed as expected", "l2SafeHeadNumber", l2SafeHead.Number)
+
+		if safeHeadNumber := l2SafeHead.Number; safeHeadNumber > 0 {
+			env.RunFaultProofProgram(t, safeHeadNumber, testCfg.CheckResult, testCfg.InputParams...)
+		}
```

Comment:
Unconditional also?

Maybe don't assume the output 0 or non-zero, but use the testcase input to determine whether or not to run the FP?